src/plugins/discover/public/context_awareness/profile_service.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+/* eslint-disable max-classes-per-file */
+
+import type { ComposableProfile, PartialProfile } from './composable_profile';
+import type { Profile } from './types';
+
+export type ResolveProfileResult<TContext> =
+  | { isMatch: true; context: TContext }
+  | { isMatch: false };
+
+export type ProfileProviderMode = 'sync' | 'async';
+
+export interface ProfileProvider<
+  TProfile extends PartialProfile,
+  TParams,
+  TContext,
+  TMode extends ProfileProviderMode
+> {
+  order: number;
+  profile: ComposableProfile<TProfile>;
+  resolve: (
+    params: TParams
+  ) => TMode extends 'sync'
+    ? ResolveProfileResult<TContext>
+    : ResolveProfileResult<TContext> | Promise<ResolveProfileResult<TContext>>;
+}
+
+abstract class BaseProfileService<
+  TProfile extends PartialProfile,
+  TParams,
+  TContext,
+  TMode extends ProfileProviderMode
+> {
+  protected readonly providers: Array<ProfileProvider<TProfile, TParams, TContext, TMode>> = [];
+
+  public registerProvider(provider: ProfileProvider<TProfile, TParams, TContext, TMode>) {
+    this.providers.push(provider);
+    this.providers.sort((a, b) => a.order - b.order);
+  }
+
+  public abstract resolve(
+    params: TParams
+  ): TMode extends 'sync' ? ComposableProfile<Profile> : Promise<ComposableProfile<Profile>>;
+}
+
+const EMPTY_PROFILE = {};
+
+export class ProfileService<
```

Comment:
What would be the downside of only using async resolution?

Reply:
Mainly that we're resolving document context in a loop over hundreds to thousands of documents, so allowing async resolution would create a lot of promises and potentially be very slow/costly depending on the implementations.

Reply:
valid point if we assume the runtime can't optimize an `async` function that never `await`s anything truly async :+1: